### PR TITLE
Add Round Spark function

### PR DESCRIPTION
**Pull request by zhejiangxiaomai**
std::round will get 0.57 for round(0.575,2).
But in  spark, we should get 0.58 in default mode.

The main reason is that double type "0.575 * 100" is actually "57.499999999999996". So we need use nextafter function to get real value.

### CODE REVIEW

**Comment by zhejiangxiaomai** [comment]
Hi, @Yuhta  would you review this PR?

**Review by rui-mo — COMMENTED** [review]
Do we need to change presto's one back?

**Comment by zhejiangxiaomai** [comment]
Yes, yuhta is backing out presto's change. We already discussed yesterday. PR is here https://github.com/facebookincubator/velox/pull/5991

File: velox/functions/sparksql/Arithmetic.h [hunk 1]
```diff
@@ -25,6 +25,43 @@
...
+    }
+
+    double factor = std::pow(10, decimals);
+    static const TNum kInf = std::numeric_limits<TNum>::infinity();
+    if (number < 0) {
+      return (std::round(std::nextafter(number, -kInf) * factor * -1) /
+              factor) *
+          -1;
+    }
+    return std::round(std::nextafter(number, kInf) * factor) / factor;
```
Comment (Yuhta) [inline velox/functions/sparksql/Arithmetic.h:55]: This probably won't work with something like `round(0.49999999999999994, 0)`.  Let's do it correctly by
1. `sprinf` to a `char[]`
2. Move the decimal point to correct position
3. `sscanf` back

**Comment by kgpai** [comment]
Hi @zhejiangxiaomai , can you rebase with latest main and push to this PR, thanks ?

File: velox/functions/sparksql/Arithmetic.h (hunk 1 above)
Comment (zhejiangxiaomai) [reply velox/functions/sparksql/Arithmetic.h:55]: Thanks, I will try this way.

**Comment by zhejiangxiaomai** [comment]
> Hi @zhejiangxiaomai , can you rebase with latest main and push to this PR, thanks ?

Just rebased.

File: velox/functions/sparksql/Arithmetic.h (hunk 1 above)
Comment (zhejiangxiaomai) [reply velox/functions/sparksql/Arithmetic.h:55]: Hi @Yuhta , I create a example code, but result is not expected. 0.575 will be convert to 0.57499999999999996 If we use sprintf. 
```
#include <iostream>
#include <cmath>
#include <limits>
using namespace std;
int main(){
 char buffer[40];
 double num = 0.49999999999999994;
 sprintf(buffer,"%.17f",num);
 string sNum = buffer;
 std::cout << "sNum : " << sNum << std::endl;

 double num1 = 0.575;
 char buffer1[40];
 sprintf(buffer1,"%.17f",num1);
 string sNum1 = buffer1;
 std::cout << "sNum1 : " << sNum1 << std::endl;
 return 0;
}

```
image

Actually, I think cases like 0.575 are quite common. and This situation `0.49999999999999994`
is very rare.

File: velox/functions/sparksql/Arithmetic.h [hunk 2]
```diff
@@ -25,6 +25,43 @@
 
 namespace facebook::velox::functions::sparksql {
 
+template <typename T>
+struct RoundFunction {
```
Comment (kgpai) [inline velox/functions/sparksql/Arithmetic.h:29]: @zhejiangxiaomai Doesn't this need to be registered ? Wont calling round in eval call the previous default implementation ?

File: velox/functions/sparksql/Arithmetic.h (hunk 1 above)
Comment (Yuhta) [reply velox/functions/sparksql/Arithmetic.h:55]: You need to check how Spark implement it and do the same then.  Both `round(0.49999999999999994, 0)` and `round(0.575, 2)` has to be correct and one is not rarer than the other.

File: velox/functions/sparksql/tests/ArithmeticTest.cpp [hunk 3]
```diff
@@ -21,6 +21,100 @@
...
+            {1.0 / 3, 6, 0.333333},   {-1.122112, 0, -1},
+            {-1.129, 1, -1.1},        {-1.129, 2, -1.13},
+            {-1.129, 2, -1.13},       {-1.0 / 3, 0, 0.0},
+            {-1.0 / 3, 1, -0.3},      {-1.0 / 3, 2, -0.33},
+            {-1.0 / 3, 6, -0.333333}, {1.0, -1, 0.0},
+            {0.0, -2, 0.0},           {-1.0, -3, 0.0},
+            {11111.0, -1, 11110.0},   {11111.0, -2, 11100.0},
+            {11111.0, -3, 11000.0},   {11111.0, -4, 10000.0},
+            {0.575, 2, 0.58},         {0.574, 2, 0.57},
+            {-0.575, 2, -0.58},       {-0.574, 2, -0.57}};
```
Comment (Yuhta) [inline velox/functions/sparksql/tests/ArithmeticTest.cpp:79]: Can you add `round(0.49999999999999994, 0)` here as well?

**Review by Yuhta — CHANGES_REQUESTED** [review]

File: velox/functions/sparksql/Arithmetic.h (hunk 2 above)
Comment (zhejiangxiaomai) [reply velox/functions/sparksql/Arithmetic.h:29]: Yes, I will fix it.

**Comment by zhejiangxiaomai** [comment]
After discussed with yuhta offline.  we think we can use `decimal_round` in gluten. And I found postgres also use "Numeric"( close to decimal) type to implement double/float round. https://github.com/postgres/postgres/blob/REL9_0_STABLE/src/backend/utils/adt/numeric.c#L827-L865

And spark also use decimal type to handle it. https://github.com/apache/spark/blob/branch-3.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1341-L1370

And https://github.com/facebookincubator/velox/pull/5597 already implement decimal_round.

**Comment by zhejiangxiaomai** [comment]
Just Remainder:
I found `nextafter` can handle `0.574999999999996`. because double is 15 decimal digits precision. `0.49999999999999994` can not get correct result because it is 17 digits.

File: velox/functions/sparksql/Arithmetic.h (hunk 2 above)
Comment (zhejiangxiaomai) [reply velox/functions/sparksql/Arithmetic.h:29]: https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/RegisterArithmetic.cpp#L49-L57 I found round were registered.

Comment (kgpai) [reply velox/functions/sparksql/Arithmetic.h:29]: Can you make sure that it calls your implementation and not presto's https://github.com/facebookincubator/velox/blob/4b653edf45cf4dafec8ef7d79e1a39063a0e6dce/velox/functions/prestosql/Arithmetic.h#L136 .

**Comment by rui-mo** [comment]
To use decimal_round, I assume we need to cast double as decimal first. But for number like `0.57499999999999996` which has 17 igits, if precision drift happens on this double value during casting process, possibly, unmatched decimal could be got and the round result is affected.

**Comment by Yuhta** [comment]
@rui-mo The change to decimal should be something at SQL parser level, and Velox should be called with decimal already.  If Velox received double, there is no way for us to recover the original value.

File: velox/functions/sparksql/Arithmetic.h (hunk 2 above)
Comment (zhejiangxiaomai) [reply velox/functions/sparksql/Arithmetic.h:29]: Thanks, I just add namespace prefix `sparksql::`.

**Comment by Yuhta** [comment]
Discussed with @zhejiangxiaomai offline, this is not solvable inside Velox so close this for now

**Pull request closed without merging**